### PR TITLE
feat(core): Validate fns signature on code creation

### LINKED ISSUE #3100: Validate entrypoints signature
### Problem to Solve

Validate entry points in `Code::try_new` as currently wrong entry points signatures are rejected only in `gear-backend-wasmi` and allowed in `gear-backend-sandbox`

### Possible Solution

.

### Notes

_No response_

### PR DESCRIPTION
**Pull request by mqxf**
Resolves #3100.

@reviewer-or-team

### CODE REVIEW

**Comment by StackOverflowExcept1on** [comment]
btw, we already have test for this, just remove `ignore`
https://github.com/gear-tech/gear/blob/37fba1c8c956b5056b42b95fe0ce2cd64ce1cf49/pallets/gear/src/tests.rs#L13756-L13765

**Comment by breathx** [comment]
@mqxf stale PR?

File: core/src/code.rs [hunk 1]
```diff
@@ -69,6 +77,34 @@ fn get_exports(
...
+
+        let types = module
+            .type_section()
+            .ok_or(CodeError::TypeSectionNotFound)?
+            .types();
+
+        let import_count = module
+            .import_section()
+            .ok_or(CodeError::ImportSectionNotFound)?
+            .entries()
```
Comment (SabaunT) [inline core/src/code.rs:94]: just use `ImportSection::functions` method, not `entries`.

File: core/src/code.rs [hunk 2]
```diff
@@ -54,13 +56,18 @@ fn get_exports(
...
     for entry in module
         .export_section()
         .ok_or(CodeError::ExportSectionNotFound)?
         .entries()
         .iter()
     {
-        if let Internal::Function(_) = entry.internal() {
+        if let Internal::Function(i) = entry.internal() {
+            if reject_unnecessary {
+                raw_exports.push(*i);
```
[the remark this answers is not in the document]
Comment (SabaunT) [reply core/src/code.rs:70]: Agree.
It's better to check it here, instead of allocating `raw_exports`.

File: core/src/code.rs [hunk 3]
```diff
@@ -54,13 +57,18 @@ fn get_exports(
 ) -> Result<BTreeSet<DispatchKind>, CodeError> {
     let mut exports = BTreeSet::<DispatchKind>::new();
 
+    let mut raw_exports = Vec::new();
```
Comment (SabaunT) [inline core/src/code.rs:60]: I'm not sure about naming as this is a vector of export functions type indexes

File: core/src/code.rs (hunk 1 above)
Comment (mqxf) [reply core/src/code.rs:94]: Wait nevermind that is exactly what I need, sorry.

**Review by gshep — APPROVED** [review]
lgtm

**Review by StackOverflowExcept1on — APPROVED** [review]

**Review by breathx — APPROVED** [review]

File: core/src/code.rs [hunk 4]
```diff
@@ -54,13 +56,35 @@ fn get_exports(
...
         .export_section()
         .ok_or(CodeError::ExportSectionNotFound)?
         .entries()
         .iter()
     {
-        if let Internal::Function(_) = entry.internal() {
+        if let Internal::Function(i) = entry.internal() {
+            if reject_unnecessary {
+                let type_id = funcs[*i as usize - import_count].type_ref();
+                let Type::Function(ref f) = types[type_id as usize];
```
Comment (techraed) [inline core/src/code.rs:83]: I'd mention why these accesses by index won't panic

**Review by techraed — APPROVED** [review]

**Pull request merged**